aya/src/programs/mod.rs

```diff
@@ -395,10 +400,209 @@ impl Drop for Program {
+                ))
+            }
+            bpf_prog_type::BPF_PROG_TYPE_LWT_SEG6LOCAL => {
+                return Err(ProgramError::IncompleteProgramDefinition(
+                    "not implemented.".to_string(),
+                ))
+            }
+            bpf_prog_type::BPF_PROG_TYPE_LIRC_MODE2 => Program::LircMode2(LircMode2 {
+                data: ProgramData::from_bpf_prog_info(name, fd, info.0)?,
+            }),
+            bpf_prog_type::BPF_PROG_TYPE_SK_REUSEPORT => {
+                return Err(ProgramError::IncompleteProgramDefinition(
+                    "not implemented.".to_string(),
+                ))
+            }
+            bpf_prog_type::BPF_PROG_TYPE_FLOW_DISSECTOR => {
+                return Err(ProgramError::IncompleteProgramDefinition(
+                    "not implemented.".to_string(),
+                ))
+            }
+            bpf_prog_type::BPF_PROG_TYPE_CGROUP_SYSCTL => Program::CgroupSysctl(CgroupSysctl {
+                data: ProgramData::from_bpf_prog_info(name, fd, info.0)?,
+            }),
+            bpf_prog_type::BPF_PROG_TYPE_RAW_TRACEPOINT_WRITABLE => {
+                return Err(ProgramError::IncompleteProgramDefinition(
+                    "not implemented.".to_string(),
+                ))
+            }
+            bpf_prog_type::BPF_PROG_TYPE_CGROUP_SOCKOPT => {
+                return Err(ProgramError::IncompleteProgramDefinition(
+                    "unable to determine attach type".to_string(),
+                ))
+            }
+            bpf_prog_type::BPF_PROG_TYPE_TRACING => {
+                return Err(ProgramError::IncompleteProgramDefinition(
+                    "unable to distinguish between fentry and fexit.".to_string(),
+                ))
+            }
+            bpf_prog_type::BPF_PROG_TYPE_STRUCT_OPS => {
+                return Err(ProgramError::IncompleteProgramDefinition(
+                    "not implemented.".to_string(),
+                ))
+            }
+            bpf_prog_type::BPF_PROG_TYPE_EXT => Program::Extension(Extension {
+                data: ProgramData::from_bpf_prog_info(name, fd, info.0)?,
+            }),
+            bpf_prog_type::BPF_PROG_TYPE_LSM => Program::Lsm(Lsm {
+                data: ProgramData::from_bpf_prog_info(name, fd, info.0)?,
+            }),
+            bpf_prog_type::BPF_PROG_TYPE_SK_LOOKUP => Program::SkLookup(SkLookup {
+                data: ProgramData::from_bpf_prog_info(name, fd, info.0)?,
+            }),
+            bpf_prog_type::BPF_PROG_TYPE_SYSCALL => {
+                return Err(ProgramError::IncompleteProgramDefinition(
+                    "not implemented.".to_string(),
+                ))
+            }
+        };
+        Ok(PinnedProgram {
+            path: PathBuf::from(path.as_ref()),
+            inner: p,
+        })
+    }
+
+    /// Removes the pinned program from bpffs.
+    pub fn unpin(&self) -> Result<(), io::Error> {
```

Comment:
@alessandrod I really want to have this as:
```rust
pub fn unpin(self) -> Result<Program, io::Error> {
    fs::remove_file(self.path.clone())?;
    Ok(self.inner)
}
```

What stopped me was that something is calling `Program::drop` :thinking: 
Even wrapping the inner `Program` in an `Option<Program>` and calling `take()` on it doesn't seem to help.